From 381c6fb12e2d6d359109d71450dc8834f8012b51 Mon Sep 17 00:00:00 2001
From: Christian Hesse <mail@eworm.de>
Date: Tue, 17 Jan 2023 22:04:08 +0100
Subject: [PATCH 03/10] FROM ARCHLINUX: properly handle i3wm
X-Developer-Signature: v=1; a=openpgp-sha256; l=4133; i=xtexchooser@duck.com;
 h=from:subject; bh=LYkM8o3M7V18ksmpo4zn78ITssSlp/KctJax0QsPJds=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDGkvVLbISRz88T6+rZ3l81Yd2TyN4gcMG792Z79PubXwb
 e2X1WniHaUsDGJcDLJiiixFhg3erDrp/KLLymVh5rAygQxh4OIUgIl4v2ZkWF3F/qp50jqrXymr
 c2a5clUbvIs+NO9u2/lMMdaZM3RWPGBk2CrRqqL1nr0gfOn+nw3HVt4+/PNR0Yqrlj/q5/5uytz
 VzAEA
X-Developer-Key: i=xtexchooser@duck.com; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

---
 .../VirtualBox/src/globals/UIDesktopWidgetWatchdog.cpp       | 3 ++-
 .../Frontends/VirtualBox/src/platform/nix/VBoxUtils-nix.cpp  | 3 +++
 .../Frontends/VirtualBox/src/platform/nix/VBoxUtils-nix.h    | 5 +++--
 src/VBox/Frontends/VirtualBox/src/widgets/UIMiniToolBar.cpp  | 3 +++
 4 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/src/VBox/Frontends/VirtualBox/src/globals/UIDesktopWidgetWatchdog.cpp b/src/VBox/Frontends/VirtualBox/src/globals/UIDesktopWidgetWatchdog.cpp
index e9b78fde4395..17a1dfc0a2c5 100644
--- a/src/VBox/Frontends/VirtualBox/src/globals/UIDesktopWidgetWatchdog.cpp
+++ b/src/VBox/Frontends/VirtualBox/src/globals/UIDesktopWidgetWatchdog.cpp
@@ -380,7 +380,8 @@ QRect UIDesktopWidgetWatchdog::availableGeometry(QScreen *pScreen) const
     /* Get cached available-geometry: */
     const QRect availableGeometry = m_availableGeometryData.value(screenToIndex(pScreen));
     /* Return cached available-geometry if it's valid or screen-geometry otherwise: */
-    return availableGeometry.isValid() ? availableGeometry : screenGeometry(pScreen);
+    return availableGeometry.isValid() && NativeWindowSubsystem::X11WindowManagerType() != X11WMType_i3 ?
+           availableGeometry : screenGeometry(pScreen);
 # endif /* !VBOX_GUI_WITH_CUSTOMIZATIONS1 */
 #else /* !VBOX_WS_NIX */
     /* Just return screen available-geometry: */
diff --git a/src/VBox/Frontends/VirtualBox/src/platform/nix/VBoxUtils-nix.cpp b/src/VBox/Frontends/VirtualBox/src/platform/nix/VBoxUtils-nix.cpp
index ab520e51b899..f168050dfdc6 100644
--- a/src/VBox/Frontends/VirtualBox/src/platform/nix/VBoxUtils-nix.cpp
+++ b/src/VBox/Frontends/VirtualBox/src/platform/nix/VBoxUtils-nix.cpp
@@ -139,6 +139,9 @@ X11WMType NativeWindowSubsystem::X11WindowManagerType()
                 else
                 if (QString((const char*)pcData).contains("Xfwm4", Qt::CaseInsensitive))
                     wmType = X11WMType_Xfwm4;
+                else
+                if (QString((const char*)pcData).contains("i3", Qt::CaseInsensitive))
+                    wmType = X11WMType_i3;
                 if (pcData)
                     XFree(pcData);
             }
diff --git a/src/VBox/Frontends/VirtualBox/src/platform/nix/VBoxUtils-nix.h b/src/VBox/Frontends/VirtualBox/src/platform/nix/VBoxUtils-nix.h
index e1dfc07f1483..0723f7598c6c 100644
--- a/src/VBox/Frontends/VirtualBox/src/platform/nix/VBoxUtils-nix.h
+++ b/src/VBox/Frontends/VirtualBox/src/platform/nix/VBoxUtils-nix.h
@@ -51,6 +51,7 @@ enum X11WMType
     X11WMType_Metacity,
     X11WMType_Mutter,
     X11WMType_Xfwm4,
+    X11WMType_i3,
 };
 
 /** X11: Screen-saver inhibit methods. */
@@ -104,9 +105,9 @@ namespace NativeWindowSubsystem
     /** Wrapper for window manager type functions. */
     X11WMType windowManagerType(bool fIsXServerAvailable);
     /** X11: Determines and returns current Window Manager type. */
-    X11WMType X11WindowManagerType();
+    SHARED_LIBRARY_STUFF X11WMType X11WindowManagerType();
     /** Wayland: Determines and returns current Window Manager type. */
-    X11WMType WaylandWindowManagerType();
+    SHARED_LIBRARY_STUFF X11WMType WaylandWindowManagerType();
 
     /** Wrapper for X11CheckExtension and WaylandCheckExtension functions. */
     bool checkExtension(bool fIsXServerAvailable, const char *extensionName);
diff --git a/src/VBox/Frontends/VirtualBox/src/widgets/UIMiniToolBar.cpp b/src/VBox/Frontends/VirtualBox/src/widgets/UIMiniToolBar.cpp
index 272729a42c22..6346fb634b1e 100644
--- a/src/VBox/Frontends/VirtualBox/src/widgets/UIMiniToolBar.cpp
+++ b/src/VBox/Frontends/VirtualBox/src/widgets/UIMiniToolBar.cpp
@@ -1076,6 +1076,9 @@ void UIMiniToolBar::prepare()
     /* Enable translucency through Qt API if supported: */
     if (uiCommon().isCompositingManagerRunning())
         setAttribute(Qt::WA_TranslucentBackground);
+
+    if (NativeWindowSubsystem::X11WindowManagerType() == X11WMType_i3)
+        setWindowFlags(Qt::FramelessWindowHint);
 #endif /* VBOX_WS_NIX */
 
     /* Make sure we have no focus: */
-- 
2.46.0

